### ISSUE TITLE
azure-provider should download intermediate certs from microsoft

### ISSUE THREAD

**Issue by djmitche** [taskcluster/taskcluster#3669]
From https://bugzilla.mozilla.org/show_bug.cgi?id=1607922
---
This bug concerns verifying worker identity based on the attested data signature that we use for registering workers. The signature is a PKCS#7 message, which is a bundle that has message, signature, and an embedded certificate. The embedded certificate is signed by an intermediate certificate, which is in turn signed by a trusted root certificate.

The initial way we're handling intermediate certs is to embed them alongside the code and loading them on startup. These certs expire in ~4 years, so we should instead download intermediate certs from Microsoft based on the authorityInfoAccess certificate extension in the embedded certificate.

RFCs 3280, 4325, 5280 have some helpful information on authorityInfoAccess.

**Comment by djmitche** [comment]
As noted in #3638, there's a good chance that Microsoft will be changing these certs soon, so let's get in front of that issue by automatically downloading them.

**Comment by jwhitlock** [comment]
I'm getting familiar with the code and the issue. The important bits in the current code are:

*  Intermediate certificates, checked into the repo at [services/worker-manager/src/providers/azure-ca-certs/embedded](https://github.com/taskcluster/taskcluster/tree/main/services/worker-manager/src/providers/azure-ca-certs).
* [AzureProvider.setup()](https://github.com/taskcluster/taskcluster/blob/main/services/worker-manager/src/providers/azure.js#L82), which [loads the certs into .caStore](https://github.com/taskcluster/taskcluster/blob/85519352cf5fd21100e5f4afdcdb06407587086e/services/worker-manager/src/providers/azure.js#L117-L119). This uses [X.509 functions](https://github.com/digitalbazaar/forge#x509) from the [forge](https://github.com/digitalbazaar/forge) library.
* [AzureProvider.registerWorker(worker, workerPool, workerIdentityProof)](https://github.com/taskcluster/taskcluster/blob/85519352cf5fd21100e5f4afdcdb06407587086e/services/worker-manager/src/providers/azure.js#L278), which processes the ``workerIdentifyProof`` in several steps:
    - Uses ``forge``'s [ASN.1 decoder](https://github.com/digitalbazaar/forge#asn) and [PKCS#7 message extractor](https://github.com/digitalbazaar/forge#pkcs7) to extract the message content, the certificate used to encode it, and the signature.
    - Verifies the signature against the the public key from the certificate, using [crypto.createVerify](https://nodejs.org/api/crypto.html#crypto_crypto_createverify_algorithm_options) from node's ``crypto`` library.
    - Checks the certificate against the ``caStore`` of intermediate certs using the X.509 function ``forge.pki.verifyCertificateChain``
    - The message payload is then processed, which includes further verification steps.

As @djmitche says in #3638, this uses the intermediate certificates, and don't include the root CA from Baltimore CyberTrust Root. The change will be that some certs may be signed by additional root CAs in the future (Baltimore CyberTrust Root and [5 others](https://docs.microsoft.com/en-us/azure/security/fundamentals/tls-certificate-changes#what-is-changing)), and we may not get a published list of intermediate certs. We may need to download an intermediate certificate at run time (``registerWorker``) to test the certificate chain.

My first stab at the needed change:

* Store the 6 root CAs in the repo, as well as known intermediate certs.
* In ``AzureProvider.setup()``, load the root CAs. For each, check if they have been revoked, and load the valid ones into ``.caStore``.
    - Alternatively, see if we can use the host OS CA store and skip checking for revocation.
    - Alternatively, rely on other processes to tell us about revoked root CA certs.
* In ``AzureProvider.setup()``, load the intermediate certs. For each, check that the issuer is already in ``.caStore``, check if they have been revoked, and add the valid ones into ``.caStore``.
    - Alternatively, rely on other processes to tell us about revoked intermediate certs.
* In ``AzureProvider.registerWorker()``, check if the issuer of the signing certificate is already in ``.caStore`` (because it was a known intermediate cert), using ``caStore.getIssuer``. If not, use the ``authorityInfoAccess`` to fetch the intermediate cert, failing if not available (Microsoft sometimes calls this the "[AIA Extension](https://docs.microsoft.com/en-us/windows-server/networking/core-network-guide/cncg/server-certs/configure-the-cdp-and-aia-extensions-on-ca1)"). Check if it has been revoked, then add it to ``.caStore``. Log that we downloaded an intermediate cert.
    - Alternatively, assume the intermediate is valid if we can fetch it.
* In ``AzureProvider.registerWorker()``, verify the certificate chain against the ``.caStore``, which should now have both the intermediate and root certificate.
* Check logs periodically to learn about new intermediate certs that we could pre-cache in the repo.

My research areas are:
* Figure out how to get access to a running azure-provider
* Get a copy of a ``workerIdentityProof`` certificate, so I can verify it has the ``authorityInfoAccess`` field, and see if it is a simple URL or a template

The implementation steps are:
* Create a first implementation with the Baltimore CyberTrust Root cert and the currently known intermediate certs, and no download ability. This _should_ work as well as the current implementation.
* Add in revocation checking
* Add in download of intermediate certs
* Remove a known intermediate, and verify the download and logging

Here's some documents mentioned here or in related earlier issues:

* [Azure TLS certificate changes](https://docs.microsoft.com/en-us/azure/security/fundamentals/tls-certificate-changes), 10/01/2020, 3 minute read
* [Microsoft PKI repository](https://www.microsoft.com/pki/mscorp/cps/default.htm)
* RFCs mentioned by @milescrabill in [bug 1607922](https://bugzilla.mozilla.org/show_bug.cgi?id=1607922)
  - [RFC 3280](https://tools.ietf.org/html/rfc3280) - Internet X.509 Public Key Infrastructure Certificate and Certificate Revocation List (CRL) Profile, April 2002, 129 pages, obsoleted by [5280](https://tools.ietf.org/html/rfc5280)
  - [RFC 4325](https://tools.ietf.org/html/rfc4325) - Internet X.509 Public Key Infrastructure Authority Information Access Certificate Revocation List (CRL) Extension, December 2005, 7 pages, obsoleted by [5280](https://tools.ietf.org/html/rfc5280)
  - [RFC 5280](https://tools.ietf.org/html/rfc5280) - Internet X.509 Public Key Infrastructure Certificate and Certificate Revocation List (CRL) Profile, May 2008, 151 pages, updated by [6818](https://tools.ietf.org/html/rfc6818), [8398](https://tools.ietf.org/html/rfc8398), and [8399](https://tools.ietf.org/html/rfc8399).
  - [RFC 6818](https://tools.ietf.org/html/rfc6818) - Updates to the Internet X.509 Public Key Infrastructure Certificate and Certificate Revocation List (CRL) Profile, January 2013, 8 pages
  - [RFC 8398](https://tools.ietf.org/html/rfc8398) - Internationalized Email Addresses in X.509 Certificates, May 2018, 12 pages
  - [RFC 8399](https://tools.ietf.org/html/rfc8399) - Internationalization Updates to [RFC 5280](https://tools.ietf.org/html/rfc5280), May 2018, 9 pages

**Comment by djmitche** [comment]
Nice work!  I think we can rely on the host for root CAs -- that's what we're doing now.  So, that should simplify things a little.  It also means it's a little safer to download intermediates based on the AIA, since they will not work if they are not signed by a real root CA.

If you just create an Azure VM manually, you should be able to use the metadata service to download the attested documents, without having to run a full worker-manager instance.

**Comment by jwhitlock** [comment]
I'd prefer to rely on the host for root CAs as well, but it is not clear to me that ``.caStore`` does that at the moment. It seems as likely that it is an empty store and the intermediate certs are trusted as pseudo-root certs. I'd like to confirm that host root CAs are used, and if they aren't, see if we can. That would vastly simplify and improve the implementation, and get us out of checking for revoked root certs.

I'm looking at the [azure worker-runner](https://github.com/taskcluster/taskcluster/tree/main/tools/worker-runner/provider/azure) code, and it does seem like I could do a lot of the research without a worker-manager instance.

**Comment by djmitche** [comment]
Oh, that's a good point regarding the roots.  Ugh.  Azure really doesn't make this easy!

**Comment by jwhitlock** [comment]
My research went in a bunch of directions since the last check-in, so this one has sections 😢 . I think I'm caught up to where @djmitche was when he wrote the issue, and might be ready to tackle some Go next week.

TL;DR:
* We can use node's CA list, and follow their security updates to keep it up-to-date.
* No changes are needed to ``azure-worker``, all the changes will be in the ``worker-manager``
* Azure includes the "Authority Information Access" extension in the attested document message, so we can use it to download the intermediate certificate, optionally check it for revocation, and verify it versus root certs.

## Root CAs

I find it impossible in just about every project to use the system certificate store. My experience is that every tool either has none (bring your own), or a baked-in set of root CAs. I guess it is too fiddly to work with the system store across platforms. **It appears ``forge`` takes the "bring your own" strategy, and ``node.js`` has a baked-in set that we can use.**

This [stack overflow answer](https://stackoverflow.com/a/21007173/10612) led me to the Node baked-in CA set in [node_root_certs.h](https://github.com/nodejs/node/blob/master/src/node_root_certs.h). That contains over 138 certs. I checked that they contain the 6 Root CAs mentioned on [Azure TLS certificate changes](https://docs.microsoft.com/en-us/azure/security/fundamentals/tls-certificate-changes). For those with ``openssl`` experience, the checking process was as awful as you imagine: manually convert to individual PEM files, run ``openssl x509 -subject -fingerprint``, and mentally convert ``D4:DE:20:..`` fingerprint format to Microsoft's ``d4de20...`` thumbprint format.

These are assigned to a local variable``root_certs`` in [crypto_context.cc](https://github.com/nodejs/node/blob/1fd7d8e90387a83fe3a2566bdac147c1b33cdbcc/src/crypto/crypto_context.cc#L43), and available with [tls.rootCertificates](https://nodejs.org/api/tls.html#tls_tls_rootcertificates) in PEM format. We could use this to populate the CA Store from ``forge``, which [defaults to empty](https://github.com/digitalbazaar/forge/blob/588c41062d9a13f8dc91be3723b159c6cc434b15/lib/x509.js#L2748-L2750) and [only loads the certs you pass to it](https://github.com/digitalbazaar/forge/blob/588c41062d9a13f8dc91be3723b159c6cc434b15/lib/x509.js#L2920-L2927). We'd then have the baked-in list of ``node.js`` certs, and rely on their security process to remove any revoked certs.

## azure-worker
I went into the ``worker-manager`` implementation above, and I peeked at the worker implementation at [tools/worker-runner/provider/azure](https://github.com/taskcluster/taskcluster/tree/main/tools/worker-runner/provider/azure). **No changes to azure-worker are needed for this issue, it's all in ``worker-manager``**. Here's the details.

* [metadata.go](https://github.com/taskcluster/taskcluster/blob/main/tools/worker-runner/provider/azure/metadata.go) has the functions to access the [metadata service](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service) at link-local address IP 169.254.169.254. I'm familiar with a similar mechanism for the [AWS metadata service](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html), so this feels familar. Azure has a few requirements for the fetch, like a ``Metadata: true`` header and an API version as a querystring parameter, like ``?api-version=2019-08-15``, handled by ``realMetadataService.fetch()``. Methods are defined for the Helpers are defined for several metadata endpoints. [queryAttestedDocument](https://github.com/taskcluster/taskcluster/blob/b92eac44f16a9a318f6398fd02dc21987cb3f26d/tools/worker-runner/provider/azure/metadata.go#L109-L123) is the one I'm interested in.
* [metadata_test.go](https://github.com/taskcluster/taskcluster/blob/main/tools/worker-runner/provider/azure/metadata_test.go) has some real-looking data. The ``attested/document`` data has the right format, but the ``signature`` is fake data, not a PKCS-#7 instance. A "real" instance is 3K, so would fill up the test file and maybe could be used in a replay attack, at least for a while.
* [azure.go](https://github.com/taskcluster/taskcluster/blob/main/tools/worker-runner/provider/azure/azure.go) has the details, [shoving the attested document](https://github.com/taskcluster/taskcluster/blob/b92eac44f16a9a318f6398fd02dc21987cb3f26d/tools/worker-runner/provider/azure/azure.go#L92-L94) into the ``workerIdentityProof``.
* [azure-test.go](https://github.com/taskcluster/taskcluster/blob/main/tools/worker-runner/provider/azure/azure_test.go) tests this, and has a more useful version of the ``/instance`` metadata than ``metadata_test.go``.

## Azure live data
Thanks to @coop and @imbstack, I was able to provision a VM and test out the metadata APIs. I provisioned a Linux VM, Ubuntu 18.04, Standard B1ls (1 vcpu, 0.5 GiB memory, super cheap). **I was able to grab an attested document, and confirmed it has the information needed to download the intermediate signing certificate.**

The instance has ``curl``and ``jq``, can be tested with:

```
curl -s -H Metadata:true "http://169.254.169.254/metadata/instance?api-version=2019-08-15" | jq
curl -s -H Metadata:true "http://169.254.169.254/metadata/scheduledevents?api-version=2017-11-01" | jq
curl -s -H Metadata:true "http://169.254.169.254/metadata/attested/document?api-version=2019-08-15" | jq
```

For ``/instance``, the [Instance API](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service#instance-api), the examples suggested calling ``/instance/compute`` and subfolders, so I wasn't sure if ``/instance`` would work. It does work, with a nested JSON structure.

My copy of ``/scheduledevents`` was empty, but otherwise like the tests.

For ``/attested/document``, I got a real PKCS-#7 message in base 64. I've been playing around with it locally with ``openssl`` after manually reformatting as PEM (``openssl pkcs7 -in message.pkcs7 -print``). The issuer was available in the AIA section:

```
            Authority Information Access: 
                CA Issuers - URI:http://www.microsoft.com/pki/mscorp/Microsoft%20IT%20TLS%20CA%205.crt
                OCSP - URI:http://ocsp.msocsp.com
```

This is the certificate that we already have checked in. I don't see why we couldn't code downloading the certificate, and maybe even use [OSCP](https://en.wikipedia.org/wiki/Online_Certificate_Status_Protocol) to double-check.

**Comment by djmitche** [comment]
That's a fantastic, detailed summary 👍🏽 

If handling possible revocations via OCSP is difficult, it's probably OK to push it out to another issue.  If a MS intermediate certificate is compromised, it'll be a bit of an apocalypse and we won't be the only ones exposed.

**Comment by jwhitlock** [comment]
I've started work in a branch, https://github.com/jwhitlock/taskcluster/tree/3669-download-azure-signing-certs, and hit a few problems. I can open a draft PR if it makes it easier to read the code. I expect there will be some helpful suggestions to improve my amateur Javascript.

[forge](https://github.com/digitalbazaar/forge), our library for dealing with certificates, doesn't load all of Node's root certificates, and only 4 of the 6 Azure root CAs load. The missing certs are "Microsoft RSA Root Certificate Authority 2017" and "Microsoft ECC Root Certificate Authority 2017". It's been known since April 2014 that ECC is not supported, which explains why the ECC cert doesn't load. The other one _should_ load, but doesn't. I am doubtful that forge will fix this soon.

I'm currently loading all the node.js root certificates (or at least, the 117 of 138 that forge can load). We may want to restrict that to the ones mentioned in the Azure document. I've added a test that the 4 Azure root CAs are loaded, and the 2 problematic ones are missing.

I continue to load the intermediate certificates, but I verify that the issuers are in the root CAs. Next, I'll add code to dynamically download a missing intermediate certificate. At that point, we can delete the certs from the repo, or retain them as an optimization.

**Comment by djmitche** [comment]
A draft PR would help us (me?) read along, yes.

So the loading problem is not a format issue, but unsupported algorithms?  That's especially icky :(

**Comment by jwhitlock** [comment]
After more investigation, the 2 missing certs appear to be loaded, but not matched by my "thumbprint" algorithm. I'll dig in and see what the issue is.

I exported the certificates that fail to load, and then inspected them with openssl (``openssl x509 -noout -text -certopt ca_default,no_serial,no_sigdump,no_extensions -fingerprint``). The Microsoft certs are not in the failed subjects:

```
Subject: C=CH, O=WISeKey, OU=OISTE Foundation Endorsed, CN=OISTE WISeKey Global Root GC CA
Subject: C=GB, ST=Greater Manchester, L=Salford, O=COMODO CA Limited, CN=COMODO ECC Certification Authority
Subject: C=GR, L=Athens, O=Hellenic Academic and Research Institutions Cert. Authority, CN=Hellenic Academic and Research Institutions ECC RootCA 2015
Subject: C=IN, OU=emSign PKI, O=eMudhra Technologies Limited, CN=emSign ECC Root CA - G3
Subject: C=US, O=AffirmTrust, CN=AffirmTrust Premium ECC
Subject: C=US, O=Amazon, CN=Amazon Root CA 3
Subject: C=US, O=Amazon, CN=Amazon Root CA 4
Subject: C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Assured ID Root G3
Subject: C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Global Root G3
Subject: C=US, O=Entrust, Inc., OU=See www.entrust.net/legal-terms, OU=(c) 2012 Entrust, Inc. - for authorized use only, CN=Entrust Root Certification Authority - EC1
Subject: C=US, O=GeoTrust Inc., OU=(c) 2007 GeoTrust Inc. - For authorized use only, CN=GeoTrust Primary Certification Authority - G2
Subject: C=US, O=Google Trust Services LLC, CN=GTS Root R3
Subject: C=US, O=Google Trust Services LLC, CN=GTS Root R4
Subject: C=US, O=VeriSign, Inc., OU=VeriSign Trust Network, OU=(c) 2007 VeriSign, Inc. - For authorized use only, CN=VeriSign Class 3 Public Primary Certification Authority - G4
Subject: C=US, O=thawte, Inc., OU=(c) 2007 thawte, Inc. - For authorized use only, CN=thawte Primary Root CA - G2
Subject: C=US, OU=emSign PKI, O=eMudhra Inc, CN=emSign ECC Root CA - C3
Subject: C=US, ST=New Jersey, L=Jersey City, O=The USERTRUST Network, CN=USERTrust ECC Certification Authority
Subject: C=US, ST=Texas, L=Houston, O=SSL Corporation, CN=SSL.com EV Root Certification Authority ECC
Subject: C=US, ST=Texas, L=Houston, O=SSL Corporation, CN=SSL.com Root Certification Authority ECC
Subject: OU=GlobalSign ECC Root CA - R4, O=GlobalSign, CN=GlobalSign
Subject: OU=GlobalSign ECC Root CA - R5, O=GlobalSign, CN=GlobalSign
```

Out of these, 5 use ``ecdsa-with-SHA256`` and 16 use ``ecdsa-with-SHA384`` as the signature algorithm.

I did find the Microsoft certs when I analyzed all the node.js certs:

```
root_cert_134
subject= /C=US/O=Microsoft Corporation/CN=Microsoft ECC Root Certificate Authority 2017
SHA1 Fingerprint=99:9A:64:C3:7F:F4:7D:9F:AB:95:F1:47:69:89:14:60:EE:C4:C3:C5

root_cert_135
subject= /C=US/O=Microsoft Corporation/CN=Microsoft RSA Root Certificate Authority 2017
SHA1 Fingerprint=73:A5:E6:4A:3B:FF:83:16:FF:0E:DC:CC:61:8A:90:6E:4E:AE:4D:74
```

So, the most likely cause is the thumbprint / fingerprint algorithm. node-forge doesn't provide one, so I used one I found on https://github.com/digitalbazaar/forge/issues/596.

**Comment by jwhitlock** [comment]
The thumbprint algorithm appears to be correct, the problem was node v12 vs node v15.

Previously, I checked [node_root_certs.h](https://github.com/nodejs/node/blob/master/src/node_root_certs.h) from node master, split the PEM-encoded strings into files, and verified that the Azure root certificates are in the list. This is true for node v15 and later, which include the same file as the master branch.

However, node v12.19.0, the version I'm using for development, has an older version of this root CA list, which does not include the two Microsoft certs.

I'm checking them into the repository for my PR. I'm loading the RSA certificate, and the thumbprint algorithm matches. The ECC certificate can't be loaded by node-forge. So, we may still need to switch libraries.

**Comment by jwhitlock** [comment]
PR #3922 is merged. It has these features:

* Continues checking against the 4 Microsoft IT certificates, so existing VMs would still be approved
* Downloads certs as needed, with additional checks, so we have a chance of handling the transition to new certs
* Checks that the attested data message is ``metadata.azure.com`` or a subdomain

There is no schedule for updating the Azure deployments, so it may be a while before we see this fix exercised in production.

Follow-on tasks:

* A security review, with changes in a follow-on PR.
* Determine when to expect the Azure certificate changes to impact the attested data API. I've asked for more details on the [Microsoft forums](https://docs.microsoft.com/answers/comments/166156/view.html).
* Issue #3922 - Check intermedia certificates for revocation. We know that the "Microsoft IT TLS CA" certs will be revoked in the future. This will strengthen the chain-of-trust process.
* Issue #3923 - Add support for ECC / ECDSA certificates. This will be needed to support the root certificate "Microsoft ECC Root Certificate Authority 2017". [node-forge](https://github.com/digitalbazaar/forge) does not include ECC, so it is unable to load the Microsoft ECC certificate, or 21 of the 138 certificates bundled with Node.js.
* Issue #3924 - Remove bundled root CAs after upgrading node. These are included in node v15, and may be backported to a node v12 release.
* Issue #3925 - Remove or repurpose legacy Microsoft intermediate certificates when revoked. These should be removed from the trusted certificate store, but may still be useful for testing.
* And any follow-on issues when the code goes live.
